### PR TITLE
[java] New rule: FinalParameterInAbstractMethod

### PR DESCRIPTION
**Pull request by vgalloy**
**Name:** FinalParameterInAbstractMethod
**Category:** Codestyle
**Description:** Declaring a method parameter as final for an interface method is useless because the implementation may choose to not respect it.

Example:

```java
public interface MyInterface {
  void process(final Object arg); // Avoid using final here
}
```

## Describe the PR

Add the rule to check that public method in interface do not use final qualifier which is unclear

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset changes 0 violations,<br />
introduces 7 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/9092915a55094b999e55f0e6ef527dd8/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by oowekyala — COMMENTED** [review]
Hi and thanks for the contribution! I'd like to take a moment to think about the scope of this rule. The problem you're trying to report is that final method parameters for interface methods are useless. This is so because the implementation doesn't have to repeat the final modifier. However this is also true for any abstract method (and there is no reason why interface methods should be singled out).

Here are my suggestions:
- The rule should flag any abstract method regardless of whether it is in an interface or not (in the xpath, just consider all methods that have no body)
- The rule should be named `FinalParameterInAbstractMethod`: it's not the rule name's job to say that "it's unclear", there are better ways to explain the purpose of the rule and the rule name doesn't have enough real estate for this. Please explain this in the description instead.
- The rule message should neutrally point out the problem: "Final parameter in abstract method" is a good message IMO.
- The rule should be placed in the code style category, as it's about a coding convention

You can also refer to our [official guidelines](https://pmd.github.io/latest/pmd_devdocs_major_rule_guidelines.html#dosdonts-rule-rules) about rule messages and rule naming

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/codestyle/FinalParameterInAbstractMethod.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule.codestyle;
+
+import net.sourceforge.pmd.testframework.PmdRuleTst;
+
+public class FinalParameterInAbstractMethod extends PmdRuleTst {
```
Comment (oowekyala) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/codestyle/FinalParameterInAbstractMethod.java:9]: This should be named `FinalParameterInAbstractMethodTest` (Test at the end)

**Review by oowekyala — APPROVED** [review]
Thanks for making these changes! This looks good to me

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/codestyle/FinalParameterInAbstractMethod.java (hunk 1 above)
Comment (vgalloy) [reply pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/codestyle/FinalParameterInAbstractMethod.java]: Arg ... wrong copy past. Sorry

**Pull request merged**